### ISSUE TITLE
NetStandard 2.1 Update?

### ISSUE THREAD

**Issue by Ryan2065** [PowerShell/PowerShellEditorServices#1297]
I just wanted to see what the plan was for updating this to .netstandard 2.1. 

I've built an Entity Framework module that loads some newer dlls built for .netstandard 2.1 and they will not load if this module has been loaded (so inside VSCode).

I've downgraded my module to use a version of Entity Framework that was released about a year ago, and can't really upgrade until PS Editor Services upgrades.

**Comment by SeeminglyScience** [comment]
Can you share the error message you are getting? What version of PowerShell are you running?

I ask because the version of netstandard that this project targets shouldn't necessarily affect how your assembly loads.  All of the dependencies that PSES load are done in a different assembly load context.  Any other dependencies loaded that would change when targeting netstandard2.1 would probably be base class libraries which PowerShell itself would load.

**Comment by Ryan2065** [comment]
I've been working off of PowerShell on Windows. The error only happens on v5.1 through the VS Code ISE window, it is fine on v7. The error I get is:

```
Method not found:
'Microsoft.Extensions.DependencyInjection.ServiceProvider Microsoft.Extensions.DependencyInjection.ServiceCollectionContainerBuil
derExtensions.BuildServiceProvider(Microsoft.Extensions.DependencyInjection.IServiceCollection)'.
```

I can "break" any PowerShell v5.1 window by simply importing this module then trying to import mine, it'll start erroring out.

You say that you are loading in a separate app domain, yet if I run this command in PowerShell v5.1 ISE I show your assemblies loaded in the main domain:

``` PowerShell
[AppDomain]::CurrentDomain.GetAssemblies()
```

Output is:

```
GAC    Version        Location
---    -------        --------
True   v4.0.30319     C:\Windows\Microsoft.NET\Framework64\v4.0.30319\mscorlib.dll
True   v4.0.30319     C:\WINDOWS\Microsoft.Net\assembly\GAC_MSIL\Microsoft.PowerShell.ConsoleHost\v4.0_3.0.0.0__31bf3856ad364e35\Mic... 
True   v4.0.30319     C:\WINDOWS\Microsoft.Net\assembly\GAC_MSIL\System\v4.0_4.0.0.0__b77a5c561934e089\System.dll
True   v4.0.30319     C:\WINDOWS\Microsoft.Net\assembly\GAC_MSIL\System.Core\v4.0_4.0.0.0__b77a5c561934e089\System.Core.dll
True   v4.0.30319     C:\WINDOWS\Microsoft.Net\assembly\GAC_MSIL\System.Management.Automation\v4.0_3.0.0.0__31bf3856ad364e35\System.... 
True   v4.0.30319     C:\WINDOWS\Microsoft.Net\assembly\GAC_MSIL\Microsoft.Management.Infrastructure\v4.0_1.0.0.0__31bf3856ad364e35\... 
True   v4.0.30319     C:\WINDOWS\Microsoft.Net\assembly\GAC_MSIL\System.Management\v4.0_4.0.0.0__b03f5f7f11d50a3a\System.Management.dll 
True   v4.0.30319     C:\WINDOWS\Microsoft.Net\assembly\GAC_MSIL\System.DirectoryServices\v4.0_4.0.0.0__b03f5f7f11d50a3a\System.Dire... 
True   v4.0.30319     C:\WINDOWS\Microsoft.Net\assembly\GAC_MSIL\System.Xml\v4.0_4.0.0.0__b77a5c561934e089\System.Xml.dll
True   v4.0.30319     C:\WINDOWS\Microsoft.Net\assembly\GAC_MSIL\System.Numerics\v4.0_4.0.0.0__b77a5c561934e089\System.Numerics.dll     
True   v4.0.30319     C:\WINDOWS\Microsoft.Net\assembly\GAC_64\System.Data\v4.0_4.0.0.0__b77a5c561934e089\System.Data.dll
True   v4.0.30319     C:\WINDOWS\Microsoft.Net\assembly\GAC_MSIL\System.Configuration\v4.0_4.0.0.0__b03f5f7f11d50a3a\System.Configur...
False  v4.0.30319
True   v4.0.30319     C:\WINDOWS\Microsoft.Net\assembly\GAC_MSIL\Microsoft.PowerShell.Security\v4.0_3.0.0.0__31bf3856ad364e35\Micros...
True   v4.0.30319     C:\WINDOWS\Microsoft.Net\assembly\GAC_64\System.Transactions\v4.0_4.0.0.0__b77a5c561934e089\System.Transaction... 
False  v4.0.30319     C:\Users\Ryan2\.vscode-insiders\extensions\ms-vscode.powershell-preview-2020.5.0\modules\PowerShellEditorServi... 
True   v4.0.30319     C:\WINDOWS\Microsoft.Net\assembly\GAC_MSIL\System.ValueTuple\v4.0_4.0.0.0__cc7b13ffcd2ddd51\System.ValueTuple.dll 
True   v4.0.30319     C:\WINDOWS\Microsoft.Net\assembly\GAC_MSIL\Microsoft.CSharp\v4.0_4.0.0.0__b03f5f7f11d50a3a\Microsoft.CSharp.dll  
True   v4.0.30319     C:\WINDOWS\Microsoft.Net\assembly\GAC_MSIL\System.Dynamic\v4.0_4.0.0.0__b03f5f7f11d50a3a\System.Dynamic.dll       
False  v4.0.30319     C:\Users\Ryan2\.vscode-insiders\extensions\ms-vscode.powershell-preview-2020.5.0\modules\PowerShellEditorServi...
True   v4.0.30319     C:\WINDOWS\Microsoft.Net\assembly\GAC_MSIL\netstandard\v4.0_2.0.0.0__cc7b13ffcd2ddd51\netstandard.dll
True   v4.0.30319     C:\WINDOWS\Microsoft.Net\assembly\GAC_MSIL\Microsoft.PowerShell.Commands.Diagnostics\v4.0_3.0.0.0__31bf3856ad3... 
True   v4.0.30319     C:\WINDOWS\Microsoft.Net\assembly\GAC_MSIL\System.Configuration.Install\v4.0_4.0.0.0__b03f5f7f11d50a3a\System.... 
True   v4.0.30319     C:\WINDOWS\Microsoft.Net\assembly\GAC_MSIL\Microsoft.PowerShell.Commands.Utility\v4.0_3.0.0.0__31bf3856ad364e3... 
True   v4.0.30319     C:\WINDOWS\Microsoft.Net\assembly\GAC_MSIL\Microsoft.PowerShell.Commands.Management\v4.0_3.0.0.0__31bf3856ad36... 
True   v4.0.30319     C:\WINDOWS\Microsoft.Net\assembly\GAC_MSIL\Microsoft.WSMan.Management\v4.0_3.0.0.0__31bf3856ad364e35\Microsoft...
True   v4.0.30319     C:\WINDOWS\Microsoft.Net\assembly\GAC_MSIL\System.Runtime.InteropServices.RuntimeInformation\v4.0_4.0.0.0__b03... 
False  v4.0.30319     C:\Users\Ryan2\.vscode-insiders\extensions\ms-vscode.powershell-preview-2020.5.0\modules\PowerShellEditorServi... 
False  v4.0.30319     C:\Users\Ryan2\.vscode-insiders\extensions\ms-vscode.powershell-preview-2020.5.0\modules\PowerShellEditorServi... 
False  v4.0.30319     C:\Users\Ryan2\.vscode-insiders\extensions\ms-vscode.powershell-preview-2020.5.0\modules\PowerShellEditorServi... 
False  v4.0.30319     C:\Users\Ryan2\.vscode-insiders\extensions\ms-vscode.powershell-preview-2020.5.0\modules\PowerShellEditorServi... 
False  v4.0.30319     C:\Users\Ryan2\.vscode-insiders\extensions\ms-vscode.powershell-preview-2020.5.0\modules\PowerShellEditorServi... 
False  v4.0.30319     C:\Users\Ryan2\.vscode-insiders\extensions\ms-vscode.powershell-preview-2020.5.0\modules\PowerShellEditorServi... 
True   v4.0.30319     C:\WINDOWS\Microsoft.Net\assembly\GAC_MSIL\System.Runtime\v4.0_4.0.0.0__b03f5f7f11d50a3a\System.Runtime.dll      
True   v4.0.30319     C:\WINDOWS\Microsoft.Net\assembly\GAC_MSIL\System.Text.Encoding\v4.0_4.0.0.0__b03f5f7f11d50a3a\System.Text.Enc...
True   v4.0.30319     C:\WINDOWS\Microsoft.Net\assembly\GAC_MSIL\System.IO\v4.0_4.0.0.0__b03f5f7f11d50a3a\System.IO.dll
True   v4.0.30319     C:\WINDOWS\Microsoft.Net\assembly\GAC_MSIL\System.IO.FileSystem\v4.0_4.0.0.0__b03f5f7f11d50a3a\System.IO.FileS... 
True   v4.0.30319     C:\WINDOWS\Microsoft.Net\assembly\GAC_MSIL\System.Runtime.Extensions\v4.0_4.0.0.0__b03f5f7f11d50a3a\System.Run... 
True   v4.0.30319     C:\WINDOWS\Microsoft.Net\assembly\GAC_MSIL\System.IO.FileSystem.Primitives\v4.0_4.0.0.0__b03f5f7f11d50a3a\Syst...
True   v4.0.30319     C:\WINDOWS\Microsoft.Net\assembly\GAC_MSIL\System.Text.Encoding.Extensions\v4.0_4.0.0.0__b03f5f7f11d50a3a\Syst... 
False  v4.0.30319     C:\Users\Ryan2\.vscode-insiders\extensions\ms-vscode.powershell-preview-2020.5.0\modules\PowerShellEditorServi... 
False  v4.0.30319     C:\Users\Ryan2\.vscode-insiders\extensions\ms-vscode.powershell-preview-2020.5.0\modules\PowerShellEditorServi... 
False  v4.0.30319     C:\Users\Ryan2\.vscode-insiders\extensions\ms-vscode.powershell-preview-2020.5.0\modules\PowerShellEditorServi... 
False  v4.0.30319     C:\Users\Ryan2\.vscode-insiders\extensions\ms-vscode.powershell-preview-2020.5.0\modules\PowerShellEditorServi... 
True   v4.0.30319     C:\WINDOWS\Microsoft.Net\assembly\GAC_MSIL\System.Web.Extensions\v4.0_4.0.0.0__31bf3856ad364e35\System.Web.Ext... 
True   v4.0.30319     C:\WINDOWS\Microsoft.Net\assembly\GAC_64\System.Web\v4.0_4.0.0.0__b03f5f7f11d50a3a\System.Web.dll
False  v4.0.30319     C:\Users\Ryan2\.vscode-insiders\extensions\ms-vscode.powershell-preview-2020.5.0\modules\PowerShellEditorServi...
False  v4.0.30319     C:\Users\Ryan2\.vscode-insiders\extensions\ms-vscode.powershell-preview-2020.5.0\modules\PowerShellEditorServi... 
False  v4.0.30319     C:\Users\Ryan2\.vscode-insiders\extensions\ms-vscode.powershell-preview-2020.5.0\modules\PowerShellEditorServi... 
False  v4.0.30319     C:\Users\Ryan2\.vscode-insiders\extensions\ms-vscode.powershell-preview-2020.5.0\modules\PowerShellEditorServi... 
False  v4.0.30319     C:\Users\Ryan2\.vscode-insiders\extensions\ms-vscode.powershell-preview-2020.5.0\modules\PowerShellEditorServi...
False  v4.0.30319     C:\Users\Ryan2\.vscode-insiders\extensions\ms-vscode.powershell-preview-2020.5.0\modules\PowerShellEditorServi... 
False  v4.0.30319     C:\Users\Ryan2\.vscode-insiders\extensions\ms-vscode.powershell-preview-2020.5.0\modules\PowerShellEditorServi... 
False  v4.0.30319     C:\Users\Ryan2\.vscode-insiders\extensions\ms-vscode.powershell-preview-2020.5.0\modules\PowerShellEditorServi... 
False  v4.0.30319     C:\Users\Ryan2\.vscode-insiders\extensions\ms-vscode.powershell-preview-2020.5.0\modules\PowerShellEditorServi... 
False  v4.0.30319     C:\Users\Ryan2\.vscode-insiders\extensions\ms-vscode.powershell-preview-2020.5.0\modules\PowerShellEditorServi... 
False  v4.0.30319     C:\Users\Ryan2\.vscode-insiders\extensions\ms-vscode.powershell-preview-2020.5.0\modules\PowerShellEditorServi... 
True   v4.0.30319     C:\WINDOWS\Microsoft.Net\assembly\GAC_MSIL\System.Threading\v4.0_4.0.0.0__b03f5f7f11d50a3a\System.Threading.dll   
False  v4.0.30319
False  v4.0.30319     C:\Users\Ryan2\.vscode-insiders\extensions\ms-vscode.powershell-preview-2020.5.0\modules\PSReadLine\2.1.0\Micr...
False  v4.0.30319     C:\Users\Ryan2\.vscode-insiders\extensions\ms-vscode.powershell-preview-2020.5.0\modules\PowerShellEditorServi... 
False  v4.0.30319     C:\Users\Ryan2\.vscode-insiders\extensions\ms-vscode.powershell-preview-2020.5.0\modules\PowerShellEditorServi... 
True   v4.0.30319     C:\WINDOWS\Microsoft.Net\assembly\GAC_MSIL\System.Runtime.Serialization\v4.0_4.0.0.0__b77a5c561934e089\System.... 
False  v4.0.30319     C:\Users\Ryan2\.vscode-insiders\extensions\ms-vscode.powershell-preview-2020.5.0\modules\PSScriptAnalyzer\1.19... 
False  v4.0.30319     C:\Users\Ryan2\.vscode-insiders\extensions\ms-vscode.powershell-preview-2020.5.0\modules\PSScriptAnalyzer\1.19...
True   v4.0.30319     C:\WINDOWS\Microsoft.Net\assembly\GAC_MSIL\System.ComponentModel.Composition\v4.0_4.0.0.0__b77a5c561934e089\Sy... 
False  v4.0.30319     C:\Users\Ryan2\.vscode-insiders\extensions\ms-vscode.powershell-preview-2020.5.0\modules\PSScriptAnalyzer\1.19... 
False  v4.0.30319     C:\Users\Ryan2\.vscode-insiders\extensions\ms-vscode.powershell-preview-2020.5.0\modules\PSScriptAnalyzer\1.19... 
```

As you can see, the AppDomain shows a lot of assemblies from PowerShellEditorServices.  One of those is the .netstandard 2.0 assembly. My *assumption* is that's the problem assembly as PowerShell 7 has 2.1, but that's as assumption.

Based on the fact that I can make any PowerShell v5.1 window crash and burn by simply importing EditorServices then importing my module with Entity Framework v3+, my guess is it's this module causing the issue.

If the AppDomain has .netstandard 2.0 loaded, how in the world can I get .netstandard 2.1 loaded? EFCore 3+ specifically calls out that it requires .netstandard 2.1.

If you want I can make a branch of my project that has the issue. I have it smooth enough that you should just have to clone down and run the "Test" script (as that builds also). Let me know, I got rid of all the EFCore 3+ references but could make something tonight.

**Comment by Ryan2065** [comment]
*note above* I say ISE a lot, all references above to ISE is the VSCode ISE window called "PowerShell Integrated Console" as that's the window that loads this module.

**Comment by SeeminglyScience** [comment]
> I've been working off of PowerShell on Windows. The error only happens on v5.1 through the VS Code ISE window, it is fine on v7. The error I get is:

Does it load in a normal PowerShell 5.1 window?  I ask because no version of net framework implements netstandard2.1.  Even if it does load, you don't get compile time safety since some of the stuff added in 2.1 will never be ported.

> ```
> Method not found:
> 'Microsoft.Extensions.DependencyInjection.ServiceProvider Microsoft.Extensions.DependencyInjection.ServiceCollectionContainerBuil
> derExtensions.BuildServiceProvider(Microsoft.Extensions.DependencyInjection.IServiceCollection)'.
> ```

Looks like a difference in the Extensions package.  The assembly resolve event handler should be taking care of that though /cc @rjmholt 

> You say that you are loading in a separate app domain, yet if I run this command in PowerShell v5.1 ISE I show your assemblies loaded in the main domain:

Same app domain, different assembly load context.  Though that's a core only concept, Windows PowerShell is handled a bit differently.  I assumed you were running core because of netstandard2.1
 
> If the AppDomain has .netstandard 2.0 loaded, how in the world can I get .netstandard 2.1 loaded? EFCore 3+ specifically calls out that it requires .netstandard 2.1.

Well the netstandard dll is provided by the runtime typically.  We do ship one, but that's only loaded if the runtime doesn't provide it.  As long as you have netfx 4.7.1 installed it should it shouldn't be loading ours, but that one will also 2.0 since netfx doesn't implement 2.1.

---

It does look like there is something that needs to be worked out in this project's assembly resolution logic, but I would advise caution trying to use netstandard2.1 in Windows PowerShell.  It's only a matter of time before you run into more missing member exceptions both inside and outside of VSCode.

**Comment by Ryan2065** [comment]
> It does look like there is something that needs to be worked out in this project's assembly resolution logic, but I would advise caution trying to use netstandard2.1 in Windows PowerShell. It's only a matter of time before you run into more missing member exceptions both inside and outside of VSCode.

Before I go too far down this rabbit hole, are you saying that things built for .netstandard 2.1 will not work well inside Windows PowerShell? So for instance Entity Framework 3.0+ which was built targeting .netstandard 2.1 will not run well inside PowerShell 5.1?

**Comment by SeeminglyScience** [comment]
> Before I go too far down this rabbit hole, are you saying that things built for .netstandard 2.1 will not work well inside Windows PowerShell? So for instance Entity Framework 3.0+ which was built targeting .netstandard 2.1 will not run well inside PowerShell 5.1?

If it happens not to reference any of the API's added between netstandard2.0 and netstandard2.1 then it *might* work fine.  If that's the case though, I'm not sure why they would target 2.1.  I think you're inevitably going to run into JIT time exceptions at some point down the road.

Here's the [netstandard support graph](https://docs.microsoft.com/en-us/dotnet/standard/net-standard#net-implementation-support).  If you look at that you'll see 2.1 shows N/A for net framework with a footnote of:

> .NET Framework won't support .NET Standard 2.1 or later versions.

The short version of why is that the run time changes needed to support the additional API's (mostly byref-like/span) aren't feasible in .NET framework,

**Comment by Ryan2065** [comment]
> If it happens not to reference any of the API's added between netstandard2.0 and netstandard2.1 then it might work fine. If that's the case though, I'm not sure why they would target 2.1. I think you're inevitably going to run into JIT time exceptions at some point down the road.

Thanks for the information.  Originally when I was in the "POC" stage before I added any code and was just creating a basic DbContext, I could do EF Core 3.1 on Windows PowerShell. Now that my project is close to completion when I switched back to EF Core 3.1 it only runs on PowerShell 7. This explains a lot.

If my work allowed PowerShell 7 I'd just drop support for Windows PowerShell, but that's not the case. I think I'll continue down this EF Core 2.2.6 path for now and in the future conditionally load EF Core 3.1 vs 2.2 based on if it's PS 7 or PS 5.1. 

Thanks for the help.
